Refactor and introduce support for attribute term data handling

marierinden opened the pull request:
### Description

This pull request introduces several enhancements and new components for handling attribute term data via the API. The changes include:

- **DTOs**: Added `AttributeTermRequest`, `AttributeTermResponse`, and `AttributeTermCollectionResponse` classes for managing and encapsulating attribute term data.
- **Resource Class**: Introduced `AttributeTermResource` to handle CRUD operations and API requests related to attribute terms. Includes validations for request types and IRIs.
- **Data Mapping**: Added `AttributeTermDataMapper` to transform API responses into DTOs and vice versa, with support for single and collection responses.
- **Refinements**: 
  - Updated docblocks in `AttributeTermDataMapper` to use a more specific request DTO class for parameter typing.
  - Updated all `@throws` annotations to use fully qualified class names, ensuring clarity and maintaining consistency.

These additions and enhancements improve code structure, maintainability, and robustness when interacting with attribute term data.

marierinden committed:
Add AttributeTermRequest DTO for handling attribute terms

This class implements the RequestInterface to manage attribute term data, including name and optional IRI fields. It also provides a method to convert the object into an array representation.

marierinden committed:
Add AttributeTermCollectionResponse DTO class

Introduces a new DTO class extending AbstractCollectionResponse to handle attribute term collection responses. This includes a typed method to ensure members are of the correct type.

marierinden committed:
Add AttributeTermDataMapper for mapping API responses

Introduces `AttributeTermDataMapper` to handle transformation of API response data into DTOs and vice versa. This implementation includes support for single and collection responses, as well as request data serialization.

marierinden committed:
Add AttributeTermResource to handle attribute term requests

Introduce AttributeTermResource class with methods for CRUD operations and querying attribute terms via the API. Includes validation for request types and IRIs, ensuring consistent contract adherence. This implementation leverages ClientInterface and DataMapperInterface for API communication and data handling.

marierinden committed:
Add AttributeTermResponse DTO class

Introduced a new AttributeTermResponse class extending AbstractResponse. This class encapsulates attribute term data and provides a getter for the name property to streamline data handling.

marierinden committed:
Refactor exception annotations to use fully qualified names

Updated all `@throws` annotations to use fully qualified class names for `ClientExceptionInterface`. This ensures clarity and eliminates ambiguity when referencing the exception in the codebase.

marierinden committed:
Update parameter type in AttributeTermDataMapper docblock

    Changed the parameter type in the docblock to use a more specific request DTO class. This ensures better type clarity and improves code maintainability.

fredrik-tveraaen-apiera committed:
Add unit tests for AttributeTerm DTOs and requests

Introduce tests for AttributeTermCollectionResponse, AttributeTermResponse, and AttributeTermRequest classes. Ensure proper coverage for instance validation, readonly properties, constructors, getters, and array conversion behavior.

fredrik-tveraaen-apiera committed:
Add unit tests for AttributeTermDataMapper

This commit introduces comprehensive unit tests for `AttributeTermDataMapper`, covering its methods for mapping responses, handling collections, converting to request data, and managing invalid or edge cases. These tests ensure correctness and robustness for the data mapping functionality.

fredrik-tveraaen-apiera committed:
Add unit and integration tests for AttributeTermResource

This commit adds unit and integration tests for `AttributeTermResource` to validate its behavior in a variety of scenarios, including finding, creating, updating, and deleting attribute terms. The tests ensure proper handling of API requests, responses, and edge cases, improving the reliability of the resource implementation.

marierinden committed:
Add attribute and store handling to AttributeTerm logic

Introduce `attribute` and `store` fields in the AttributeTerm DTOs and update the data mapper to handle these fields. Modify AttributeTermResource to validate and use attribute IRIs in endpoint URLs, aligning with the updated API structure.

fredrik-tveraaen-apiera commented on src/DataMapper/AttributeTermDataMapper.php:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * @param \Apiera\Sdk\DTO\Request\Category\CategoryRequest $requestDto
```
This must be changes to correct DTO. It should be \Apiera\Sdk\DTO\Request\AttributeTerm\AttributeTermRequest $requestDto

fredrik-tveraaen-apiera commented on src/DataMapper/AttributeTermDataMapper.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        try {
+            $members = array_map(
+                fn(array $attributeTerm): ResponseInterface =>
```
ResponseInterface should be replaced with AttributeTermResponse

fredrik-tveraaen-apiera commented on src/Resource/AttributeTermResource.php:
```diff
@@ -0,0 +1,172 @@
+     * @throws \Apiera\Sdk\Interface\ClientExceptionInterface
+     * @throws InvalidRequestException
+     */
+    public function find(
```
No need to wrap this into multiple lines as it would fit within the cs of max 120 limit

marierinden committed:
Fix incorrect types in AttributeTermDataMapper methods

    Updated the return type for the array_map callback and corrected the typehint for the request DTO parameter. These changes ensure type consistency and prevent potential type-related errors.

marierinden committed:
Simplify method declaration in find function

Condensed the method signature to a single line for improved readability. No functional changes were made to the logic or behavior of the method.

fredrik-tveraaen-apiera committed:
Merge release/0.3.0: Update test files and add AttributeTerm implementation

fredrik-tveraaen-apiera committed:
Add examples for managing attribute terms in documentation

This update includes code examples for finding, searching, creating, updating, and deleting attribute terms. It enhances the documentation by providing practical usage scenarios to aid developers in implementing these features.
